Mathlib/NumberTheory/LegendreSymbol/AddCharacter.lean

```diff
@@ -56,54 +56,43 @@ lemma val_mem_rootsOfUnity (φ : AddChar R R') (a : R) (h : 0 < ringChar R) :
 
 /-- An additive character is *primitive* iff all its multiplicative shifts by nonzero
 elements are nontrivial. -/
-def IsPrimitive (ψ : AddChar R R') : Prop :=
-  ∀ a : R, a ≠ 0 → IsNontrivial (mulShift ψ a)
+def IsPrimitive (ψ : AddChar R R') : Prop := ∀ ⦃a : R⦄, a ≠ 0 → mulShift ψ a ≠ 1
 #align add_char.is_primitive AddChar.IsPrimitive
 
 /-- The composition of a primitive additive character with an injective mooid homomorphism
 is also primitive. -/
 lemma IsPrimitive.compMulHom_of_isPrimitive {R'' : Type*} [CommMonoid R''] {φ : AddChar R R'}
     {f : R' →* R''} (hφ : φ.IsPrimitive) (hf : Function.Injective f) :
-    (f.compAddChar φ).IsPrimitive := by
-  intro a a_ne_zero
-  obtain ⟨r, ne_one⟩ := hφ a a_ne_zero
-  rw [mulShift_apply] at ne_one
-  simp only [IsNontrivial, mulShift_apply, f.coe_compAddChar, Function.comp_apply]
-  exact ⟨r, fun H ↦ ne_one <| hf <| f.map_one ▸ H⟩
+    (f.compAddChar φ).IsPrimitive := fun a ha ↦ by
+  simpa [DFunLike.ext_iff] using (MonoidHom.compAddChar_injective_right f hf).ne (hφ ha)
 
 /-- The map associating to `a : R` the multiplicative shift of `ψ` by `a`
 is injective when `ψ` is primitive. -/
 theorem to_mulShift_inj_of_isPrimitive {ψ : AddChar R R'} (hψ : IsPrimitive ψ) :
     Function.Injective ψ.mulShift := by
   intro a b h
   apply_fun fun x => x * mulShift ψ (-b) at h
-  simp only [mulShift_mul, mulShift_zero, add_right_neg] at h
-  have h₂ := hψ (a + -b)
-  rw [h, isNontrivial_iff_ne_trivial, ← sub_eq_add_neg, sub_ne_zero] at h₂
-  exact not_not.mp fun h => h₂ h rfl
+  simp only [mulShift_mul, mulShift_zero, add_right_neg, mulShift_apply] at h
+  simpa [← sub_eq_add_neg, sub_eq_zero] using (hψ . h)
 #align add_char.to_mul_shift_inj_of_is_primitive AddChar.to_mulShift_inj_of_isPrimitive
 
 -- `AddCommGroup.equiv_direct_sum_zmod_of_fintype`
 -- gives the structure theorem for finite abelian groups.
 -- This could be used to show that the map above is a bijection.
 -- We leave this for a later occasion.
 /-- When `R` is a field `F`, then a nontrivial additive character is primitive -/
-theorem IsNontrivial.isPrimitive {F : Type u} [Field F] {ψ : AddChar F R'} (hψ : IsNontrivial ψ) :
-    IsPrimitive ψ := by
-  intro a ha
-  cases' hψ with x h
-  use a⁻¹ * x
-  rwa [mulShift_apply, mul_inv_cancel_left₀ ha]
-#align add_char.is_nontrivial.is_primitive AddChar.IsNontrivial.isPrimitive
+theorem IsPrimitive.of_ne_one {F : Type u} [Field F] {ψ : AddChar F R'} (hψ : ψ ≠ 1) :
+    IsPrimitive ψ :=
+  fun a ha h ↦ hψ $ by simpa [mulShift_mulShift, ha] using congr_arg (mulShift · a⁻¹) h
```

Comment:
I thought `$` was deprecated in favour of `<|` in mathlib

Reply:
This is my reading of the style guide as well: "Note: while $ is a synonym for <|, its use in mathlib is disallowed in favor of <| for consistency as well as because of the symmetry with |>."